### PR TITLE
fix ios flutter engine  memory leak

### PR DESCRIPTION
**Pull request by Natoto**
solved the issue  FlutterMethodChannel may cause memory leak in iOS 

 https://github.com/flutter/flutter/issues/26007

the demo:
https://github.com/Natoto/flutterOnExistApp/tree/flutter1.2

 the docment:
https://juejin.im/post/5c24ad306fb9a049d2361cff 

there is a big circular reference between register enginer and messger of method channel

### CODE REVIEW

**Comment by Natoto** [comment]
I signed it!

**Comment by Natoto** [comment]
> agreements

HuaNgbO  is me

**Comment by Natoto** [comment]
I signed it!

**Comment by Natoto** [comment]
I signed it!

**Comment by Natoto** [comment]
I signed it! , HuaNgbO is me,please check my CLA

**Comment by Natoto** [comment]
if the engine gets released and the channel will released first @dnfield

File: shell/platform/darwin/ios/ios_gl_render_target.mm [hunk 1]
```diff
@@ -64,7 +64,8 @@
 IOSGLRenderTarget::~IOSGLRenderTarget() {
   [EAGLContext setCurrentContext:context_];
   FML_DCHECK(glGetError() == GL_NO_ERROR);
-
+ //fix platformview memory leak https://github.com/flutter/flutter/issues/24714
+  [EAGLContext setCurrentContext:context_.get()];
```
Comment (dnfield) [inline shell/platform/darwin/ios/ios_gl_render_target.mm]: Why are we setting this twice to the same value?  It's also not clear to me that this belongs in the same PR...

File: shell/platform/darwin/ios/ios_gl_context.mm [hunk 2]
```diff
@@ -22,6 +22,7 @@
     resource_context_.reset([[EAGLContext alloc] initWithAPI:kEAGLRenderingAPIOpenGLES2
                                                   sharegroup:context_.get().sharegroup]);
   }
+  [[NSNotificationCenter defaultCenter] postNotificationName:@"EAGLContextsharegroup" object:context_.get().sharegroup];
```
Comment (dnfield) [inline shell/platform/darwin/ios/ios_gl_context.mm]: What is this for?  It probably belongs in a separate PR.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 3]
```diff
@@ -416,6 +416,8 @@ - (void)viewDidAppear:(BOOL)animated {
 
 - (void)viewWillDisappear:(BOOL)animated {
   TRACE_EVENT0("flutter", "viewWillDisappear");
+  [[[UIApplication sharedApplication] keyWindow] endEditing:YES];
```
Comment (dnfield) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm]: This probably belongs in a separate PR.

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm [hunk 4]
```diff
@@ -576,7 +587,7 @@ - (instancetype)initWithPlugin:(NSString*)pluginKey flutterEngine:(FlutterEngine
   self = [super init];
   NSAssert(self, @"Super init cannot be nil");
   _pluginKey = [pluginKey retain];
-  _flutterEngine = [flutterEngine retain];
+  _flutterEngine = flutterEngine;//[flutterEngine retain];
```
Comment (dnfield) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: Here and elsewhere: please remove the commented out code.

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm [hunk 5]
```diff
@@ -266,21 +267,31 @@ - (void)setupChannels {
 
 - (void)maybeSetupPlatformViewChannels {
   if (_shell && self.shell.IsSetup()) {
+      
+    __block FlutterPlatformPlugin *blockmethochannel = (FlutterPlatformPlugin*)_platformPlugin.get();
```
Comment (dnfield) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: can we get a more descriptive name and have camelCasing on the variable name?

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm [hunk 6]
```diff
@@ -266,21 +267,31 @@ - (void)setupChannels {
...
     [_platformChannel.get() setMethodCallHandler:^(FlutterMethodCall* call, FlutterResult result) {
-      [_platformPlugin.get() handleMethodCall:call result:result];
+      [blockmethochannel handleMethodCall:call result:result];
     }];
 
+    __block __typeof(self)weakSelf = self;
```
Comment (dnfield) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: does this not need to be marked as `__weak`?

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm [hunk 7]
```diff
@@ -266,21 +267,31 @@ - (void)setupChannels {
...
+
     [_platformViewsChannel.get()
         setMethodCallHandler:^(FlutterMethodCall* call, FlutterResult result) {
-          _platformViewsController->OnMethodCall(call, result);
+            [weakSelf platformCall:call result:result];
         }];
 
+    __block FlutterTextInputPlugin *blockinputplugin = (FlutterTextInputPlugin*)_textInputPlugin.get();
```
Comment (dnfield) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: ditto on variable name.  Maybe `textInputPlugin` would be fine?

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm (hunk 5 above)
Comment (dnfield) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: `... platformPlugin = ..;`

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm [hunk 8]
```diff
@@ -8,6 +8,7 @@
 
 #include <memory>
 
+#include "flutter/fml/memory/weak_ptr.h"
```
Comment (dnfield) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: This doesn't look like it's used.

File: shell/platform/darwin/ios/framework/Source/FlutterChannels.mm [hunk 9]
```diff
@@ -251,15 +251,15 @@ - (instancetype)initWithName:(NSString*)name
...
+  _messenger = messenger;
   _codec = [codec retain];
   return self;
 }
 
 - (void)dealloc {
   [_name release];
   [_codec release];
-  [_messenger release];
+  //[_messenger release];
```
Comment (dnfield) [inline shell/platform/darwin/ios/framework/Source/FlutterChannels.mm]: remove commented out code.

File: shell/platform/darwin/ios/framework/Source/FlutterChannels.mm [hunk 10]
```diff
@@ -172,14 +172,14 @@ - (instancetype)initWithName:(NSString*)name
...
-  _messenger = [messenger retain];
+  _messenger = messenger;
   _codec = [codec retain];
   return self;
 }
 
 - (void)dealloc {
   [_name release];
-  [_messenger release];
+  //[_messenger release];
```
Comment (dnfield) [inline shell/platform/darwin/ios/framework/Source/FlutterChannels.mm]: remove commented out code.

File: shell/platform/darwin/ios/framework/Source/FlutterChannels.mm [hunk 11]
```diff
@@ -32,14 +32,14 @@ - (instancetype)initWithName:(NSString*)name
...
```
Comment (dnfield) [inline shell/platform/darwin/ios/framework/Source/FlutterChannels.mm]: remove commented out code.

File: shell/platform/darwin/ios/framework/Source/FlutterChannels.mm [hunk 12]
```diff
@@ -32,14 +32,14 @@ - (instancetype)initWithName:(NSString*)name
   self = [super init];
   NSAssert(self, @"Super init cannot be nil");
   _name = [name retain];
-  _messenger = [messenger retain];
+  _messenger = messenger;
```
Comment (dnfield) [inline shell/platform/darwin/ios/framework/Source/FlutterChannels.mm]: Here and elsewhere: what do we do if someone has a referenc eto a message channel and is sending it messages, but there's no messenger anymore?  Log a warning?  assert?

File: shell/platform/darwin/ios/ios_gl_render_target.mm (hunk 1 above)
Comment (Natoto) [reply shell/platform/darwin/ios/ios_gl_render_target.mm]: hi, this is belongs other pr, i'll remove it next commit

File: shell/platform/darwin/ios/ios_gl_context.mm (hunk 2 above)
Comment (Natoto) [reply shell/platform/darwin/ios/ios_gl_context.mm]: hi, this is belongs other PR, i'll remove it next commit

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 3 above)
Comment (Natoto) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm]: hi, this is belongs other pr, i'll remove it next commit

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm (hunk 4 above)
Comment (Natoto) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: okay

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm (hunk 6 above)
Comment (Natoto) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: in MRC files we should use __block for weak object

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm (hunk 8 above)
Comment (Natoto) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: got it， i will remove it next commit

File: shell/platform/darwin/ios/framework/Source/FlutterChannels.mm (hunk 9 above)
Comment (Natoto) [reply shell/platform/darwin/ios/framework/Source/FlutterChannels.mm]: got it

File: shell/platform/darwin/ios/framework/Source/FlutterChannels.mm (hunk 12 above)
Comment (Natoto) [reply shell/platform/darwin/ios/framework/Source/FlutterChannels.mm]: to breake the circular reference we should not  retain messager， because the messager is holded by channel‘s caller

File: shell/platform/darwin/ios/ios_gl_render_target.mm (hunk 1 above)
Comment (Natoto) [reply shell/platform/darwin/ios/ios_gl_render_target.mm]: done

File: shell/platform/darwin/ios/ios_gl_context.mm (hunk 2 above)
Comment (Natoto) [reply shell/platform/darwin/ios/ios_gl_context.mm]: done

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 3 above)
Comment (Natoto) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm]: done

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm (hunk 4 above)
Comment (Natoto) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: done

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm (hunk 5 above)
Comment (Natoto) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: use __block  avoid cycle reference

File: shell/platform/darwin/ios/framework/Source/FlutterChannels.mm (hunk 12 above)
Comment (dnfield) [reply shell/platform/darwin/ios/framework/Source/FlutterChannels.mm]: Yes, but what happens if a caller with a reference to the channel tries to use it? This fixes the retain cycle but we could end up with zombies here, right?

File: shell/platform/darwin/ios/framework/Source/FlutterChannels.mm (hunk 10 above)
Comment (Natoto) [reply shell/platform/darwin/ios/framework/Source/FlutterChannels.mm]: done

File: shell/platform/darwin/ios/framework/Source/FlutterChannels.mm (hunk 12 above)
Comment (Natoto) [reply shell/platform/darwin/ios/framework/Source/FlutterChannels.mm]: the messager actually is flutterengine which is holded by flutterviewcontroller or created by user, if flutterviewcontroller is existed, the engine existed, could  not be zombies

Comment (dnfield) [reply shell/platform/darwin/ios/framework/Source/FlutterChannels.mm]: But someone might have grabbed a reference to a message channel from a flutter engine or viewcontroller, and then try to use it after the engine is gone.

Comment (Natoto) [reply shell/platform/darwin/ios/framework/Source/FlutterChannels.mm]: In your case, when engine is gone, the engine is nil,  nil call any method and would not work, may not be crashed. 
this is my demo https://github.com/Natoto/flutterOnExistApp/tree/multiflutter
then you can see the total memory released,when close flutter viewcontroller
can you show your case when case crash.

**Comment by dnfield** [comment]
Can you either post from your other commit, or can you fix your commit history so the CLA bot can verify this?

**Comment by dnfield** [comment]
This code will result in a crash based on the current version of the patch:

```objc
FlutterBasicMessageChannel* channel;
  @autoreleasepool {
    FlutterViewController* flutterViewController =
      [[FlutterViewController alloc] init];
    channel = flutterViewController.engine.systemChannel;
  }
  [channel sendMessage:@"Hello!"];
```

You could imagine other steps in there, such as showing and then dismissing the FlutterViewController.

File: shell/platform/darwin/ios/framework/Source/FlutterChannels.mm (hunk 12 above)
Comment (Natoto) [reply shell/platform/darwin/ios/framework/Source/FlutterChannels.mm]: > This code will result in a crash based on the current version of the patch:
> 
> ```objc
>   FlutterBasicMessageChannel* channel;
>   @autoreleasepool {
>     FlutterViewController* flutterViewController =
>       [[FlutterViewController alloc] init];
>     channel = flutterViewController.engine.systemChannel;
>   }
>   [channel sendMessage:@"Hello!"];
> ```
> 
> You could imagine other steps in there, such as showing and then dismissing the FlutterViewController. We should not crash in this instance.

a compromise solution in FlutterEngine.m  method
 
avoid access zombies object and crash ,_shell maybe not reset succ or engine may be release but not nil ,use lifecyclechannel check if engine is gone
```
- (void)sendOnChannel:(NSString*)channel
              message:(NSData*)message
          binaryReply:(FlutterBinaryReply)callback{
 ...

    if(_lifecycleChannel.get()){
        _shell->GetPlatformView()->DispatchPlatformMessage(platformMessage);
    }
```

Comment (dnfield) [reply shell/platform/darwin/ios/framework/Source/FlutterChannels.mm]: If we don't reset the shell the engine can't be released :(

**Comment by cbracken** [comment]
@Natoto do you plan to proceed with this PR?

**Comment by Natoto** [comment]
> @Natoto do you plan to proceed with this PR? If so, would you mind rebasing against `HEAD` of master?

yes, i haven't found better solution for the autorelease question , please help me rebase it

**Comment by dnfield** [comment]
@Natoto - this needs some formatting.  I'm also still concerned about the memory management around the _messenger - we need to do something to make sure we're not creating a potential crash here.  I want to see the memory leak go away, but not at the cost of a crash.  Can you address that?  Otherwise, I think this may have to be taken up in a separate PR.

**Comment by cbracken** [comment]
@Natoto will you have time to address the latest round of @dnfield 's comments? Let us know how you'd like to proceed.

**Comment by cbracken** [comment]
Thanks for your contribution. Since we haven't seen a reply, we're going to reluctantly close this PR for now, but if you get time to address the comments, we'd be more than happy to take a look at a new PR. If you do send a new PR, it would be very useful to reference this one in the description in order to help reviewers see the previous context.
